File: frontend/src/lib/types/chart-spec.ts
/**
 * Universal Chart Specification
 * Especificación compartida entre frontend y backend para generación de charts
 */

// ============================================================================
// Tipos Base
// ============================================================================

export type ChartType = 'line' | 'bar' | 'area' | 'pie' | 'd3_pyramid' | 'mapa';

export interface WeekMetadata {
  year: number;
  week: number;
  start_date: string;
  end_date: string;
}

export interface Dataset {
  label?: string | null;
  data: number[];
  color?: string | null;
  type?: 'area' | 'line' | null; // Para area charts con líneas superpuestas
}

// ============================================================================
// Datos por Tipo de Chart
// ============================================================================

export interface BaseChartData {
  labels: string[];
  datasets: Dataset[];
  metadata?: WeekMetadata[] | null;
}

export interface PyramidDataPoint {
  age_group: string;
  male: number;
  female: number;
  [key: string]: unknown;
}

export interface MapDepartmentData {
  codigo_indec: number;
  nombre: string;
  zona_ugd: string;
  poblacion: number;
  casos: number;
  tasa_incidencia: number;
}

export interface MapChartData {
  departamentos: MapDepartmentData[];
  total_casos: number;
}

// ============================================================================
// Configuración de Charts
// ============================================================================

export interface BaseChartConfig {
  height?: number | null;
  width?: number | null;
  showLegend?: boolean | null;
  showGrid?: boolean | null;
  colors?: string[] | null;
}

export interface LineChartConfig extends BaseChartConfig {
  showPoints?: boolean | null;
  curved?: boolean | null;
}

export interface BarChartConfig extends BaseChartConfig {
  stacked?: boolean | null;
  horizontal?: boolean | null;
}

export interface AreaChartConfig extends BaseChartConfig {
  stacked?: boolean | null;
  fillOpacity?: number | null;
}

export interface PieChartConfig extends BaseChartConfig {
  showPercentages?: boolean | null;
  innerRadius?: number | null; // For donut charts
}

export interface PyramidChartConfig extends BaseChartConfig {
  showAxisLabels?: boolean | null;
}

export interface MapChartConfig extends BaseChartConfig {
  colorScale?: 'sequential' | 'diverging' | null;
  province?: 'chubut' | null; // Expandible a otras provincias
}

// ============================================================================
// Filtros (para reproducibilidad)
// ============================================================================

export interface ChartFilters {
  grupo_eno_ids?: number[] | null;
  tipo_eno_ids?: number[] | null;
  clasificacion?: string[] | null;
  provincia_id?: number[] | null;
  fecha_desde?: string | null;
  fecha_hasta?: string | null;
  edad_min?: number | null;
  edad_max?: number | null;
  tipo_sujeto?: 'humano' | 'animal' | null;
  [key: string]: unknown; // Filtros adicionales
}

// ============================================================================
// Especificación Universal
// ============================================================================

export interface UniversalChartSpec {
  // Metadata
  id: string;
  title: string;
  description?: string | null;
  codigo?: string | null; // Código del chart en el backend

  // Tipo y datos (discriminated union)
  type: ChartType;
  data: ChartData;
  config: ChartConfig;

  // Filtros aplicados (para reproducibilidad)
  filters?: ChartFilters | null;

  // Timestamp de generación
  generated_at?: string | null;
}

// Discriminated union para data
export type ChartData =
  | { type: 'line' | 'bar' | 'area' | 'pie'; data: BaseChartData }
  | { type: 'd3_pyramid'; data: PyramidDataPoint[] }
  | { type: 'mapa'; data: MapChartData };

// Discriminated union para config
export type ChartConfig =
  | { type: 'line'; config: LineChartConfig }
  | { type: 'bar'; config: BarChartConfig }
  | { type: 'area'; config: AreaChartConfig }
  | { type: 'pie'; config: PieChartConfig }
  | { type: 'd3_pyramid'; config: PyramidChartConfig }
  | { type: 'mapa'; config: MapChartConfig };

// ============================================================================
// Request/Response para API
// ============================================================================

export interface ChartSpecRequest {
  chart_code: string;
  filters: ChartFilters;
  config?: Partial<ChartConfig>;
}

export interface ChartSpecResponse {
  spec: UniversalChartSpec;
  generated_at: string;
}

// ============================================================================
// Utilidades de Validación
// ============================================================================

export function isValidChartType(type: string): type is ChartType {
  return ['line', 'bar', 'area', 'pie', 'd3_pyramid', 'mapa'].includes(type);
}

export function validateChartSpec(spec: unknown): spec is UniversalChartSpec {
  if (typeof spec !== 'object' || spec === null) return false;

  const s = spec as Partial<UniversalChartSpec>;

  return !!(
    s.id &&
    s.title &&
    s.type &&
    isValidChartType(s.type) &&
    s.data &&
    s.config
  );
}

// ============================================================================
// Helpers para crear specs
// ============================================================================

export function createLineChartSpec(
  id: string,
  title: string,
  data: BaseChartData,
  config?: Partial<LineChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'line',
    data: { type: 'line', data },
    config: { type: 'line', config: { ...config } },
  };
}

export function createBarChartSpec(
  id: string,
  title: string,
  data: BaseChartData,
  config?: Partial<BarChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'bar',
    data: { type: 'bar', data },
    config: { type: 'bar', config: { ...config } },
  };
}

export function createAreaChartSpec(
  id: string,
  title: string,
  data: BaseChartData,
  config?: Partial<AreaChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'area',
    data: { type: 'area', data },
    config: { type: 'area', config: { ...config } },
  };
}

export function createPieChartSpec(
  id: string,
  title: string,
  data: BaseChartData,
  config?: Partial<PieChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'pie',
    data: { type: 'pie', data },
    config: { type: 'pie', config: { ...config } },
  };
}

export function createPyramidChartSpec(
  id: string,
  title: string,
  data: PyramidDataPoint[],
  config?: Partial<PyramidChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'd3_pyramid',
    data: { type: 'd3_pyramid', data },
    config: { type: 'd3_pyramid', config: { ...config } },
  };
}

export function createMapChartSpec(
  id: string,
  title: string,
  data: MapChartData,
  config?: Partial<MapChartConfig>
): UniversalChartSpec {
  return {
    id,
    title,
    type: 'mapa',
    data: { type: 'mapa', data },
    config: { type: 'mapa', config: { ...config } },
  };
}

File: backend/app/schemas/chart_spec.py
"""
Universal Chart Specification
Schemas Pydantic para la especificación universal de charts
Compatibles con los tipos TypeScript del frontend
"""

from typing import List, Optional, Dict, Any, Literal, Union, Annotated
from pydantic import BaseModel, Field
from datetime import datetime


# ============================================================================
# Tipos Base
# ============================================================================

ChartType = Literal["line", "bar", "area", "pie", "d3_pyramid", "mapa"]


class WeekMetadata(BaseModel):
    """Metadata de semana epidemiológica"""
    year: int
    week: int
    start_date: str
    end_date: str


class Dataset(BaseModel):
    """Dataset para charts line/bar/area/pie"""
    label: Optional[str] = None
    data: List[float]
    color: Optional[str] = None
    type: Optional[Literal["area", "line"]] = None  # Para area charts con líneas


# ============================================================================
# Datos por Tipo de Chart
# ============================================================================

class BaseChartData(BaseModel):
    """Datos para line, bar, area, pie charts"""
    labels: List[str]
    datasets: List[Dataset]
    metadata: Optional[List[WeekMetadata]] = None


class PyramidDataPoint(BaseModel):
    """Punto de datos para pirámide poblacional"""
    age_group: str
    male: float
    female: float


class MapDepartmentData(BaseModel):
    """Datos de departamento para mapa"""
    codigo_indec: int
    nombre: str
    zona_ugd: str
    poblacion: int
    casos: int
    tasa_incidencia: float


class MapChartData(BaseModel):
    """Datos para mapa de Chubut"""
    departamentos: List[MapDepartmentData]
    total_casos: int


# ============================================================================
# Discriminated Union Wrappers para Data
# ============================================================================

class LineChartData(BaseModel):
    """Wrapper con discriminador para line chart data"""
    type: Literal["line"] = "line"
    data: BaseChartData


class BarChartData(BaseModel):
    """Wrapper con discriminador para bar chart data"""
    type: Literal["bar"] = "bar"
    data: BaseChartData


class AreaChartData(BaseModel):
    """Wrapper con discriminador para area chart data"""
    type: Literal["area"] = "area"
    data: BaseChartData


class PieChartData(BaseModel):
    """Wrapper con discriminador para pie chart data"""
    type: Literal["pie"] = "pie"
    data: BaseChartData


class PyramidChartData(BaseModel):
    """Wrapper con discriminador para pyramid chart data"""
    type: Literal["d3_pyramid"] = "d3_pyramid"
    data: List[PyramidDataPoint]


class MapChartDataWrapper(BaseModel):
    """Wrapper con discriminador para map chart data"""
    type: Literal["mapa"] = "mapa"
    data: MapChartData


# Union discriminada para todos los tipos de datos
ChartDataUnion = Annotated[
    Union[
        LineChartData,
        BarChartData,
        AreaChartData,
        PieChartData,
        PyramidChartData,
        MapChartDataWrapper,
    ],
    Field(discriminator="type"),
]


# ============================================================================
# Configuración de Charts
# ============================================================================

class BaseChartConfig(BaseModel):
    """Configuración base para todos los charts"""
    height: Optional[int] = None
    width: Optional[int] = None
    showLegend: Optional[bool] = True
    showGrid: Optional[bool] = True
    colors: Optional[List[str]] = None


class LineChartConfig(BaseChartConfig):
    """Configuración específica para line charts"""
    showPoints: Optional[bool] = True
    curved: Optional[bool] = False


class BarChartConfig(BaseChartConfig):
    """Configuración específica para bar charts"""
    stacked: Optional[bool] = False
    horizontal: Optional[bool] = False


class AreaChartConfig(BaseChartConfig):
    """Configuración específica para area charts"""
    stacked: Optional[bool] = False
    fillOpacity: Optional[float] = 0.6


class PieChartConfig(BaseChartConfig):
    """Configuración específica para pie charts"""
    showPercentages: Optional[bool] = True
    innerRadius: Optional[int] = 0  # For donut charts


class PyramidChartConfig(BaseChartConfig):
    """Configuración específica para pyramid charts"""
    showAxisLabels: Optional[bool] = True


class MapChartConfig(BaseChartConfig):
    """Configuración específica para map charts"""
    colorScale: Optional[Literal["sequential", "diverging"]] = "sequential"
    province: Optional[Literal["chubut"]] = "chubut"


# ============================================================================
# Discriminated Union Wrappers para Config
# ============================================================================

class LineChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para line chart config"""
    type: Literal["line"] = "line"
    config: LineChartConfig


class BarChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para bar chart config"""
    type: Literal["bar"] = "bar"
    config: BarChartConfig


class AreaChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para area chart config"""
    type: Literal["area"] = "area"
    config: AreaChartConfig


class PieChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para pie chart config"""
    type: Literal["pie"] = "pie"
    config: PieChartConfig


class PyramidChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para pyramid chart config"""
    type: Literal["d3_pyramid"] = "d3_pyramid"
    config: PyramidChartConfig


class MapChartConfigWrapper(BaseModel):
    """Wrapper con discriminador para map chart config"""
    type: Literal["mapa"] = "mapa"
    config: MapChartConfig


# Union discriminada para todos los tipos de config
ChartConfigUnion = Annotated[
    Union[
        LineChartConfigWrapper,
        BarChartConfigWrapper,
        AreaChartConfigWrapper,
        PieChartConfigWrapper,
        PyramidChartConfigWrapper,
        MapChartConfigWrapper,
    ],
    Field(discriminator="type"),
]


# ============================================================================
# Filtros (para reproducibilidad)
# ============================================================================

class ChartFilters(BaseModel):
    """Filtros aplicados al chart (para reproducibilidad)"""
    grupo_eno_ids: Optional[List[int]] = None
    tipo_eno_ids: Optional[List[int]] = None
    clasificacion: Optional[List[str]] = None
    provincia_id: Optional[List[int]] = None
    fecha_desde: Optional[str] = None
    fecha_hasta: Optional[str] = None
    edad_min: Optional[int] = None
    edad_max: Optional[int] = None
    tipo_sujeto: Optional[Literal["humano", "animal"]] = None
    extra: Optional[Dict[str, Any]] = Field(default_factory=dict)


# ============================================================================
# Especificación Universal
# ============================================================================

class UniversalChartSpec(BaseModel):
    """
    Especificación universal de chart
    Puede ser usada tanto por frontend como backend
    """
    # Metadata
    id: str
    title: str
    description: Optional[str] = None
    codigo: Optional[str] = None

    # Tipo de chart
    type: ChartType

    # Datos (con discriminador type para TypeScript)
    data: ChartDataUnion

    # Config (con discriminador type para TypeScript)
    config: ChartConfigUnion

    # Filtros aplicados (para reproducibilidad)
    filters: Optional[ChartFilters] = None

    # Timestamp
    generated_at: Optional[str] = Field(
        default_factory=lambda: datetime.now().isoformat()
    )

    class Config:
        json_schema_extra = {
            "example": {
                "id": "chart_001",
                "title": "Casos por Semana Epidemiológica",
                "type": "line",
                "data": {
                    "labels": ["SE 1", "SE 2", "SE 3"],
                    "datasets": [
                        {"label": "Confirmados", "data": [10, 15, 12]},
                    ],
                },
                "config": {"height": 400, "showLegend": True},
            }
        }


# ============================================================================
# Request/Response para API
# ============================================================================

class ChartSpecRequest(BaseModel):
    """Request para obtener spec de un chart"""
    chart_code: str
    filters: ChartFilters
    config: Optional[Dict[str, Any]] = None


class ChartSpecResponse(BaseModel):
    """Response con el spec generado"""
    spec: UniversalChartSpec
    generated_at: str
